### PR TITLE
 cmd/telemeter-server,pkg/cache: cache auth resps

### PR DESCRIPTION
**Pull request by squat**
This commit adds new functionality to Telemeter server to enable it to
cache the responses to authentication requests made while handling
remote-write requests. The only currently implemented backend is
Memcached, because it is so simple, but any shared k-v would serve our
purposes.

cc @bwplotka @aditya-konarde @brancz @kakkoyun

### CODE REVIEW

**Comment by brancz** [comment]
/lgtm

Impressed how fast you threw this out there. Nice!

File: cmd/telemeter-server/main.go [hunk 1]
```diff
@@ -86,6 +89,7 @@ func main() {
 		PartitionKey:       "_id",
 		Ratelimit:          4*time.Minute + 30*time.Second,
 		TTL:                10 * time.Minute,
+		MemcachedExpire:    24 * 60 * 60,
```
Comment (metalmatze) [inline cmd/telemeter-server/main.go]: 24 hours seem a bit long. Are the JWTs valid for the same duration? I would have said 1 hour would be long already.

Comment (squat) [reply cmd/telemeter-server/main.go]: That’s the current expiration time of the JWTs :/ so I think we should default to this for now just to keep traffic to tollbooth somewhat steady. We can of course override this with the flag at any point.

**Comment by squat** [comment]
and totally agree about the e2e tests. Started getting into the weeds of it last night. We need:
* obsv/up generating metrics
* telemeter ingesting
* telemeter auth as a stub authorizer
* memcached for caching
* thanos receive ingesting
* thanos query for testing that metric made it through

File: pkg/authorize/tollbooth/tollbooth.go [hunk 2]
```diff
@@ -75,3 +77,12 @@ func (a *authorizer) AuthorizeCluster(token, cluster string) (string, error) {
 
 	return response.AccountID, nil
 }
+
+// ExtractToken extracts the token from an auth request.
+// In the case of a request to Tollbooth, the token
+// is the entire contents of the request body.
+func ExtractToken(r *http.Request) (string, error) {
+	body, err := ioutil.ReadAll(r.Body)
```
Comment (bwplotka) [inline pkg/authorize/tollbooth/tollbooth.go]: Leaking body, let's close it.

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		return nil, errors.Wrap(err, "failed to generate key from request")
+	}
+
+	raw, ok, err := r.c.Get(key)
+	if err != nil {
+		r.cacheReadsTotal.WithLabelValues("error").Inc()
+		return nil, errors.Wrap(err, "failed to retrieve value from cache")
+	}
+
+	if !ok {
```
Comment (bwplotka) [inline pkg/cache/cache.go]: I would vote for hit being a main quick shortcut and miss a passing to next. The rule of thumb is that easier/shorter should go first.

File: pkg/cache/cache.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		r.cacheReadsTotal.WithLabelValues("miss").Inc()
+		resp, err := r.next.RoundTrip(req)
+		if err == nil && resp.StatusCode/200 == 1 {
+			// Try to cache the response but don't block.
+			defer func() {
+				raw, err := httputil.DumpResponse(resp, true)
+				if err != nil {
+					level.Error(r.l).Log("msg", "failed to dump response", "err", err)
```
Comment (bwplotka) [inline pkg/cache/cache.go]: It's kind of cache write error, right?

File: pkg/cache/memcached/memcached.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	i := memcache.Item{
+		Key:        key,
+		Value:      value,
+		Expiration: c.expiration,
+	}
+	return c.Client.Set(&i)
+}
+
+// hashKey hashes the given key to ensure that it is less than 250 bytes,
+// as Memcached cannot handler longer keys.
```
Comment (bwplotka) [inline pkg/cache/memcached/memcached.go]:
```suggestion
// as Memcached cannot handle longer keys.
```

File: pkg/cache/cache.go (hunk 4 above)
Comment (squat) [reply pkg/cache/cache.go]: yeah kind of, but not request was made to memcached. I'd like to ensure we have metrics to track failures to contact the cache.

Comment (bwplotka) [reply pkg/cache/cache.go]: Makes sense :+1:

File: pkg/authorize/handler.go [hunk 6]
```diff
@@ -129,7 +132,8 @@ func NewHandler(logger log.Logger, client *http.Client, endpoint *url.URL, tenan
 
 		token, err := base64.StdEncoding.DecodeString(authParts[1])
 		if err != nil {
-			http.Error(w, "bad authorization header", http.StatusBadRequest)
+			level.Warn(logger).Log("msg", "failed to extract token", "err", err)
+			w.WriteHeader(http.StatusBadRequest)
```
[the remark this answers is not in the document]
Comment (squat) [reply pkg/authorize/handler.go]: :/ nope, I think it's a mistake. I'll fix it here

**Pull request merged**